### PR TITLE
Feature caching

### PR DESCRIPTION
**Pull request by reuben**
This doubles as road-paving for an easier tf.data.Dataset implementation of the feeding code.

### CODE REVIEW

File: util/feeding.py [hunk 1]
```diff
@@ -1,13 +1,109 @@
...
+    transcript = text_to_char_array(file.transcript, alphabet)
+
+    if (2*numcontext + len(features)) < len(transcript):
+        raise ValueError('Error: Audio file {} is too short for transcription.'.format(file.wav_filename))
+
+    return features, len(features), transcript, len(transcript)
+
+
+# load samples from CSV, compute features, optionally cache results on disk
+def preprocess(csv_files, batch_size, numcep, numcontext, alphabet, hdf5_cache_path=None):
```
Comment (tilmankamp) [inline util/feeding.py]: How about putting `preprocess` and its dependencies into an extra file that would have no TF dependencies and so on? Then we could also provide a simple command-line wrapper around it (under `/bin`).

File: DeepSpeech.py [hunk 2]
```diff
@@ -402,7 +406,7 @@ def BiRNN(batch_x, seq_length, dropout, reuse=False, batch_size=None, n_steps=-1
     # This is done to prepare the batch for input into the first layer which expects a tensor of rank `2`.
 
     # Permute n_steps and batch_size
-    batch_x = tf.transpose(batch_x, [1, 0, 2])
+    batch_x = tf.transpose(batch_x, [1, 0, 2, 3])
     # Reshape to prepare input for first layer
     batch_x = tf.reshape(batch_x, [-1, n_input + 2*n_input*n_context]) # (n_steps*batch_size, n_input + 2*n_input*n_context)
```
Comment (tilmankamp) [inline DeepSpeech.py]: Shouldn't your additional dimension in `batch_x = tf.transpose(batch_x, [1, 0, 2, 3])` somehow occur in this comment?

File: DeepSpeech.py [hunk 3]
```diff
@@ -40,6 +40,10 @@ def create_flags():
     tf.app.flags.DEFINE_string  ('test_files',       '',          'comma separated list of files specifying the dataset used for testing. multiple files will get merged')
     tf.app.flags.DEFINE_boolean ('fulltrace',        False,       'if full trace debug info should be generated during training')
 
+    tf.app.flags.DEFINE_string  ('train_cached_features_path',      '',          'comma separated list of files specifying the dataset used for training. multiple files will get merged')
+    tf.app.flags.DEFINE_string  ('dev_cached_features_path',        '',          'comma separated list of files specifying the dataset used for validation. multiple files will get merged')
+    tf.app.flags.DEFINE_string  ('test_cached_features_path',       '',          'comma separated list of files specifying the dataset used for testing. multiple files will get merged')
```
Comment (tilmankamp) [inline DeepSpeech.py]: Pretty long names...

**Comment by tilmankamp** [comment]
For the sake of simplicity (this PR changes the affected code): How about already solving #1537 ? Should be doable in 2-3 lines.

File: DeepSpeech.py (hunk 2 above)
Comment (reuben) [reply DeepSpeech.py]: The comment is referring to the shape *after* the reshape, which is not changed.

File: util/feeding.py (hunk 1 above)
Comment (reuben) [reply util/feeding.py]: OK.

File: DeepSpeech.py (hunk 3 above)
Comment (reuben) [reply DeepSpeech.py]: What shorter name do you suggest? I tried to come up with something smaller but couldn't think of options that still described what the flag does completely.

Comment (tilmankamp) [reply DeepSpeech.py]: It's OK to keep them like that.

**Pull request merged**